### PR TITLE
many: add configuration for splash screen

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Adds configuration to the base for splash screen. This requires also initramfs changes. Just a draft for the moment as we need a new plymouth deb package in jammy and some snapd changes.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
This is ready for an initial pass now. Note that to have the full feature the changes in the initramfs are needed, and a gadget with additional `quiet splash` kernel command line parameters is required.

File: hooks/026-configure-plymouth.chroot [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#! /bin/sh -ex
+
+# Rewrite units for plymouth
+cat << 'EOF' > /lib/systemd/system/plymouth-start.service
```
Comment (valentindavid) [inline hooks/026-configure-plymouth.chroot:4]: I feel like we should use drop-ins to tweak upstream units, that would be easier to maintain.

**Comment by valentindavid** [comment]
`plymouth-{halt,poweroff,reboot,kexec}.service` are missing `--ignore-serial-consoles`. Without it, it seems to not display anything when shutting down.

File: hooks/026-configure-plymouth.chroot (hunk 1 above)
Comment (alfonsosanchezbeato) [reply hooks/026-configure-plymouth.chroot]: Changed now to use overrides

**Comment by alfonsosanchezbeato** [comment]
> `plymouth-{halt,poweroff,reboot,kexec}.service` are missing `--ignore-serial-consoles`. Without it, it seems to not display anything when shutting down.

Added `--ignore-serial-consoles` in the places where it was not yet.

**Comment by valentindavid** [comment]
Do we need `/usr/libexec/plymouth/plymouth-generate-initrd`, `/usr/libexec/plymouth/plymouth-populate-initrd`, `/usr/libexec/plymouth/plymouth-update-initrd`?

Also I have noticed we have files in `/usr/share/initramfs-tools`.

File: static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+After=snap-initramfs-mounts.service
+Requires=plymouth-quit.service splash-client.service
+Requires=plymouth-show-run.service plymouth-show-install.service
+Requires=plymouth-show-recover.service plymouth-show-factory-reset.service
```
Comment (valentindavid) [inline static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf:5]: I would prefer `Wants`. Specially if we move `plymouth-quit.service` from `multi-user.target.wants`.

File: hooks/026-configure-plymouth.chroot [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+# Note that we have overrides for plymouth services in:
+# static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf
+# static/usr/lib/systemd/system/plymouth-quit.service.d/core-override.conf
+# static/usr/lib/systemd/system/plymouth-halt.service.d/core-override.conf
+# static/usr/lib/systemd/system/plymouth-reboot.service.d/core-override.conf
+# static/usr/lib/systemd/system/plymouth-kexec.service.d/core-override.conf
+# static/usr/lib/systemd/system/plymouth-poweroff.service.d/core-override.conf
+
+cat << 'EOF' > /lib/systemd/system/splash-client.service
```
Comment (valentindavid) [inline hooks/026-configure-plymouth.chroot:11]: Those files are not dynamic. Could we move them into `static/`?

Comment (valentindavid) [reply hooks/026-configure-plymouth.chroot:11]: Maybe we can keep the configuration files here. But the systemd units would be better in `static/`.

File: hooks/026-configure-plymouth.chroot [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ConditionKernelCommandLine=snapd_recovery_mode=run
+ConditionPathExists=/run/plymouth/pid
+
+[Service]
+ExecStart=-/bin/plymouth display-message --text='Starting system...'
+Type=oneshot
+TimeoutSec=20
+
+[Install]
+WantedBy=sysinit.target
```
Comment (valentindavid) [inline hooks/026-configure-plymouth.chroot:51]: I am not sure why there is an `[Install]` section if it there is `Requires` added in `plymouth-start.service`

File: static/usr/bin/splash-client [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash -exu
```
Comment (valentindavid) [inline static/usr/bin/splash-client:1]: Maybe the `x` can be removed.

Comment (valentindavid) [reply static/usr/bin/splash-client:1]: If you were using for debugging, I suppose.

File: static/usr/lib/systemd/system/plymouth-quit.service.d/core-override.conf [hunk 6]
```diff
@@ -0,0 +1,6 @@
+[Unit]
+After=snapd.seeded.service
+ConditionKernelCommandLine=snapd_recovery_mode=run
+
+[Service]
+ExecStartPost=-systemctl stop splash-client.service
```
Comment (valentindavid) [inline static/usr/lib/systemd/system/plymouth-quit.service.d/core-override.conf:6]: I would use `plymouth --ping` in the loop of `splash-client` instead so we do not have to change `plymouth-quit.service`.

File: static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+[Service]
+ExecStart=
+ExecStart=/usr/sbin/plymouthd --debug --mode=boot --pid-file=/run/plymouth/pid --attach-to-session --ignore-serial-consoles
```
Comment (valentindavid) [inline static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf:9]: I suppose we can remove `--debug`.

File: static/usr/bin/splash-client [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+# Change 1 is "Initialize system state"
+msg=
+prev_msg=
+while true; do
+    sleep 0.5
+    msg=$(snap change --abs-time 1 | grep -m1 Doing |
```
Comment (valentindavid) [inline static/usr/bin/splash-client:8]: Is this suppose to show messages only on the first boot?

Comment (valentindavid) [reply static/usr/bin/splash-client:8]: Would `snap tasks --last=seed? --abs-time` be more obvious?

Should we also check `--last=become-operational`?

File: static/usr/bin/splash-client [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+              awk -F ' +' '{ s = ""; for (i = 4; i <= NF; i++) s = s $i " "; print s }') \
```
Comment (valentindavid) [inline static/usr/bin/splash-client:9]: This is fine. But here are some alternatives:

Without awk:
```
tr -s " " | cut -d" " -f4-
```

Merging awk and grep together, and removing the loop (which also does not squash multiple spaces):
```
awk -F " +" '$1 == "Doing" { print substr($0, index($0,$4)); exit }'
```

Comment (valentindavid) [reply static/usr/bin/splash-client:9]: And for the lol in sed:
```
sed -E '/^Doing /{;s/([^ ]*[ ]*){3}//;q};d'
```

File: hooks/001-extra-packages.chroot [hunk 10]
```diff
@@ -19,25 +19,45 @@ trap 'umount /proc' EXIT
...
+# We need to install first the certificates so the cert for the PPA is trusted
+apt update
+apt install --no-install-recommends -y ca-certificates
+
 # shellcheck disable=SC1091
 CODENAME=$(. /etc/os-release; echo "$UBUNTU_CODENAME")
-# enable the foundations ubuntu-image PPA
-echo "deb http://ppa.launchpadcontent.net/canonical-foundations/ubuntu-image/ubuntu $CODENAME main" > /etc/apt/sources.list.d/ubuntu-image.list
+# enable the ucdev PPA with additional packages to build bases
+echo "deb https://ppa.launchpadcontent.net/ucdev/base-ppa/ubuntu/ $CODENAME main" \
```
Comment (valentindavid) [inline hooks/001-extra-packages.chroot:29]: It is because of `https` that `ca-certificates` is needed.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the review! I've addressed/answered the comments now.

File: hooks/026-configure-plymouth.chroot (hunk 3 above)
Comment (alfonsosanchezbeato) [reply hooks/026-configure-plymouth.chroot]: Done now

File: static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf (hunk 2 above)
Comment (alfonsosanchezbeato) [reply static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf]: Done

**Comment by alfonsosanchezbeato** [comment]
I've also removed some additional packages from the plymouth deb that we don't need as suggested

**Review by valentindavid — APPROVED** [review]
Looks good.

File: static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf [hunk 11]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+After=snap-initramfs-mounts.service
```
Comment (valentindavid) [inline static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf:2]: This unit does not exist in core-base, and if did exist would create the same cycle as in core-initrd.

Comment (alfonsosanchezbeato) [reply static/usr/lib/systemd/system/plymouth-start.service.d/core-override.conf]: Yeah, good point. I've removed it now.

**Review by Meulengracht — APPROVED** [review]
LGTM!

**Pull request merged**